### PR TITLE
737 adding resource without grant

### PR DESCRIPTION
**Pull request by davidsmejia**
## Issue Number

#737 

## Purpose/Implementation Notes

* adds a generic hook to talking to hubspot api
* adds a `GrantsRequired` guard that forced the feature request form if user's account does not have any grants
* `GrantsRequired` runs after `AccountRequired` so as it stands right now they would have to login to find out they cant post without a grant.
* validates the input which is handled by the hook using generic helpers (very cool) and takes an arbitrary form schema
* alerts user to success/errors and redirects them home when they are done

NOTE: if the hubspot form updates, we will need to update the form component to match expected inputs... there was no way to use the embed code in a way that would suit our needs.

## Types of changes

- New feature (non-breaking change which adds functionality)

## Functional tests

Tested validation, successful response, and bad response all seem to be working as expected

## Checklist

- [X] Lint and unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

## Screenshots

Screen Shot 2021-03-19 at 4 05 35 PM
Screen Shot 2021-03-19 at 4 06 16 PM

Updated:

Screen Shot 2021-03-19 at 4 41 15 PM

### CODE REVIEW

**Review by dvenprasad — APPROVED** [review]
Could we add. (optional) next to `Is there anything you would like to add` ?

Otherwise it looks good! 🐈 🐈

File: client/src/components/GrantRequired.js [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+  // show feature request form if user does not have a grant
+  if (!userOk) {
+    return (
+      <>
+        {children}
+        <Modal showing nondismissable>
+          <ResourceAddNoGrantRequestForm
+            portalId="5187852"
+            formId="960eaa6e-dc9a-4921-874f-40ba9bae8583"
```
Comment (kurtwheeler) [inline client/src/components/GrantRequired.js:29]: Are these magic strings from Hubspot?

**Comment by kurtwheeler** [comment]
Hmm, is it possible that a lab member could be confused by this if they are trying to list or import a resource for their PI who does have an Alex's grant? The form doesn't mention that they may just need an invite to their PI's team.

File: client/src/components/GrantRequired.js (hunk 1 above)
Comment (kurtwheeler) [reply client/src/components/GrantRequired.js:29]: Oh I see it now. Maybe leave a comment or something? Hubspot isn't mentioned in this file anywhere.

File: client/src/hooks/useUser.js [hunk 2]
```diff
@@ -110,6 +110,26 @@ export const useUser = (defaultUser, defaultToken) => {
...
+    )
+
+    if (!userIsOk) return false
+
+    // user has grants
+    const { grants, organizations: teams } = freshUser
+
+    const hasGrants = grants.length > 0
+    // join all the arrays of org grants to one array and check if empty
+    const hasTeamGrants = [].concat(...teams.map((t) => t.g)).length > 0
```
[the remark this answers is not in the document]
Comment (davidsmejia) [reply client/src/hooks/useUser.js:128]: ooph good catch

File: client/src/hooks/useHubspotForm.js [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    } catch (e) {
+      setErrors(getErrorMessages(e))
+      return false
+    }
+
+    return true
+  }
+
+  // can't just post the json data, no that would be too easy
+  // convert to {fields: [{name, value}]}
```
Comment (kurtwheeler) [inline client/src/hooks/useHubspotForm.js:44]: Hahahaha that's horrible!

File: client/src/components/GrantRequired.js (hunk 1 above)
Comment (davidsmejia) [reply client/src/components/GrantRequired.js]: ok i moved them to env vars

**Comment by davidsmejia** [comment]
> Hmm, is it possible that a lab member could be confused by this if they are trying to list or import a resource for their PI who does have an Alex's grant? The form doesn't mention that they may just need an invite to their PI's team.

@dvenprasad might be able to give a response to that... they would only see this if they do not have a grant given to them when signing up or if they are not yet a member of a team with a grant

**Review by kurtwheeler — APPROVED** [review]
LGTM but we might want to add a bit of help for someone who's just trying to list a resource for their PI.

**Pull request merged**